src/librustdoc/clean/types.rs

```diff
@@ -388,6 +388,42 @@ impl Item {
         self.def_id.as_def_id().map(|did| tcx.get_attrs(did).inner_docs()).unwrap_or(false)
     }
 
+    crate fn visibility(&self, tcx: TyCtxt<'_>) -> Visibility {
+        let mut def_id = match self.def_id {
+            // Anything but DefId *shouldn't* matter, but return a reasonable value anyway.
+            ItemId::Auto { .. } | ItemId::Blanket { .. } => return Visibility::Inherited,
+            ItemId::Primitive(..) => return Visibility::Public,
+            ItemId::DefId(did) => did,
+        };
+        match *self.kind {
+            // Variant fields inherit their enum's visibility.
+            StructFieldItem(..) if is_field_vis_inherited(tcx, def_id) => {
+                return Visibility::Inherited;
+            }
+            // Variants always inherit visibility
+            VariantItem(..) => return Visibility::Inherited,
+            // Return the visibility of `extern crate` statement, not the crate itself (which will always be public).
+            ExternCrateItem { crate_stmt_id, .. } => def_id = crate_stmt_id.to_def_id(),
+            // Trait items inherit the trait's visibility
+            AssocConstItem(..) | AssocTypeItem(..) | TyMethodItem(..) | MethodItem(..) => {
+                let is_trait_item = match tcx.associated_item(def_id).container {
+                    ty::TraitContainer(_) => true,
+                    ty::ImplContainer(impl_id) => tcx.impl_trait_ref(impl_id).is_some(),
+                };
+                if is_trait_item {
+                    return Visibility::Inherited;
+                }
+            }
+            _ => {}
+        }
+        // If this item was inlined, show the visibility of the `use` statement, not the definition.
```

Comment:
This doesn't seem quite correct. Take this code for example:

```rust
// crate a

pub struct S {
  pub x: i32,
  y: i32,
}

// crate b

pub(crate) use a::S;
```

If `--document-private-items` is passed, then I think the documentation for `b::S` **should** look like this (although I may not be remembering the inlining rules quite right):

```rust
pub(crate) struct S {
  pub(crate) x: i32,
  y: i32,
}
```

But with your change, I think it will look like this:

```rust
pub(crate) struct S {
  pub(crate) x: i32,
  pub(crate) y: i32,
```

Can you test that code (or a variant of it to get the inlining to happen) before and after this change to check?

Reply:
Or perhaps this won't happen because inlining doesn't occur for struct *fields*? But I feel like this sort of bug could probably occur with other items.

Reply:
This is exactly the same behavior as before, whatever that happened to be. I just moved the logic from `try_inline` to `visibility()`.

Reply:
also, I don't think this can happen for struct fields because you can't import them directly, only the struct itself. Enum variants can be imported, but they're special-cased earlier anyway.

Reply:
I couldn't get `y` to appear with `pub(crate) use` (maybe due to https://github.com/rust-lang/rust/pull/91094 ?) but with `pub use` the docs look like this on nightly:
![image](https://user-images.githubusercontent.com/23638587/144327690-dce80280-a474-47bf-980a-2d791d8b43c4.png)
so uhh yeah this is a pre-existing bug, good catch :sweat_smile:

Reply:
> This is exactly the same behavior as before, whatever that happened to be. I just moved the logic from try_inline to visibility().

Ah, I see.